### PR TITLE
feat: adds init async with timeout and deprecated non-timeout init functions

### PR DESCRIPTION
**Pull request by tanderson-ld**
**Requirements**

- [x] I have added test coverage for new or changed functionality
- [x] I have followed the repository's [pull request submission guidelines](../blob/main/CONTRIBUTING.md#submitting-pull-requests)
- [x] I have validated my changes against all supported platform versions

**Related issues**

242110

### CODE REVIEW

**Comment by tanderson-ld** [comment]
TODO: Add test for init with timeout.

File: src/LaunchDarkly.ClientSdk/Interfaces/ILdClient.cs [hunk 1]
```diff
@@ -349,7 +348,6 @@ public interface ILdClient : IDisposable
         /// about setting the context and optionally requesting a unique key for it</param>
         /// <returns>a task that yields true if new flag values were obtained</returns>
         /// <seealso cref="Identify(Context, TimeSpan)"/>
-        /// <seealso cref="ILdClientExtensions.IdentifyAsync(ILdClient, User)"/>
```
Comment (tanderson-ld) [inline src/LaunchDarkly.ClientSdk/Interfaces/ILdClient.cs:352]: For reviewers: User is deprecated

File: src/LaunchDarkly.ClientSdk/LdClient.cs [hunk 2]
```diff
@@ -244,16 +281,19 @@ async Task StartAsync()
...
         /// <para>
-        /// In offline mode, this constructor will return immediately. Otherwise, it will wait and block on
-        /// the current thread until initialization and the first response from the LaunchDarkly service is
-        /// returned, up to the specified timeout. If the timeout elapses, the returned instance will have
-        /// an <see cref="Initialized"/> property of <see langword="false"/>.
+        /// The constructor will return the <see cref="LdClient"/> instance once the first response from
+        /// the LaunchDarkly service is returned, or immediately if offline, or when the the specified
+        /// wait time elapses.  If the max wait time elapses, the returned instance will have
+        /// an <see cref="Initialized"/> property of <see langword="false"/>, but the instance will continue
+        /// trying to get fresh feature flags.
```
Comment (tanderson-ld) [inline src/LaunchDarkly.ClientSdk/LdClient.cs:288]: For reviewers: Documentation changes are just tidying up the contents and making each of the init docs more consistent with each other.  They diverged in subtle ways over time.

File: src/LaunchDarkly.ClientSdk/LdClient.cs [hunk 3]
```diff
@@ -317,7 +358,53 @@ async Task StartAsync()
...
+        /// device characteristics including manufacturer, model, operating system, locale, and so on. We recommend
+        /// enabling this when you configure the SDK.  See
+        /// <a href="https://docs.launchdarkly.com/sdk/features/environment-attributes">our documentation</a> for
+        /// more details.</param>
+        /// <param name="initialContext">the initial evaluation context; see <see cref="LdClient"/> for more
+        /// about setting the context and optionally requesting a unique key for it</param>
+        /// <returns>a Task that resolves to the singleton LdClient instance</returns>
+        [Obsolete("Initializing the LDClient without a timeout is no longer permitted to help prevent" +
+                  "consumers from blocking their application execution by mistake when connectivity is poor.  Please" +
+                  "use InitAsync(string, ConfigurationBuilder.AutoEnvAttributes, Context, TimeSpan) and specify a max wait time.")]
```
Comment (tanderson-ld) [inline src/LaunchDarkly.ClientSdk/LdClient.cs:389]: For reviewers: The diff engine makes this look like I added a method and deprecated it immediately.  That is not the case, just the diff shifting things around.

File: src/LaunchDarkly.ClientSdk/LdClient.cs [hunk 4]
```diff
@@ -226,14 +234,43 @@ public sealed class LdClient : ILdClient
...
             if (!success)
             {
-                _log.Warn("Client did not successfully initialize within {0} milliseconds.",
-                    maxWaitTime.TotalMilliseconds);
+                _log.Warn(DidNotInitializeTimelyWarning, maxWaitTime.TotalMilliseconds);
             }
         }
 
+        /// <summary>
+        /// Starts the client and waits up to the wait time to initialize feature flags.  If offline,
```
Comment (cwaldren-ld) [inline src/LaunchDarkly.ClientSdk/LdClient.cs:250]: Should it mention that initialization will continue in the background, even if the timeout elapses?

File: src/LaunchDarkly.ClientSdk/LdClient.cs [hunk 5]
```diff
@@ -226,14 +234,43 @@ public sealed class LdClient : ILdClient
...
+        {
+            if (maxWaitTime >= ExcessiveInitWaitTime)
+            {
+                _log.Warn(
+                    ExcessiveInitWaitTimeWarning,
+                    maxWaitTime, ExcessiveInitWaitTime);
+            }
+
+            var startTask = _connectionManager.Start();
+            var completedTask = await Task.WhenAny(startTask, Task.Delay(maxWaitTime));
```
Comment (cwaldren-ld) [inline src/LaunchDarkly.ClientSdk/LdClient.cs:264]: How does this `!=` logic work? (just haven't seen it before with tasks)

File: src/LaunchDarkly.ClientSdk/LdClient.cs [hunk 6]
```diff
@@ -463,13 +555,48 @@ public static LdClient Init(Configuration config, Context initialContext, TimeSp
...
+                  "use InitAsync(Configuration, Context, TimeSpan) and specify a max wait time.")]
         public static async Task<LdClient> InitAsync(Configuration config, Context initialContext)
         {
             var c = CreateInstance(config, initialContext, TimeSpan.Zero);
             await c.StartAsync();
             return c;
         }
 
+        /// <summary>
+        /// Creates a new <see cref="LdClient"/> singleton instance and attempts to initialize feature flags
```
Comment (cwaldren-ld) [inline src/LaunchDarkly.ClientSdk/LdClient.cs:569]: What is the word "singleton" doing for us here?

File: src/LaunchDarkly.ClientSdk/LdClient.cs (hunk 5 above)
Comment (tanderson-ld) [reply src/LaunchDarkly.ClientSdk/LdClient.cs:264]: The whenAny returns the task that completed and if it isn't the start task then it must be the delay task.

File: src/LaunchDarkly.ClientSdk/LdClient.cs (hunk 6 above)
Comment (tanderson-ld) [reply src/LaunchDarkly.ClientSdk/LdClient.cs:569]: Helps customer's realize that they don't make multiple of these.

Comment (tanderson-ld) [reply src/LaunchDarkly.ClientSdk/LdClient.cs:569]: Further down is 

        /// You must use one of these static factory methods to instantiate the single instance of LdClient
        /// for the lifetime of your application.

**Review by cwaldren-ld — APPROVED** [review]

**Pull request merged**